flocker/provision/_libcloud.py

```diff
@@ -341,6 +344,270 @@ def _cleanup_node_named(self, name):
+                        ),
+                        name=name,
+                        id=node.id,
+                        state=node.state,
+                    ).write()
+                    return _retry_exception_async(reactor, node.destroy)
+
+        d.addCallback(destroy_if_found)
+        return d.result
+
+    def _async_refresh_node(self, reactor, target_node):
+        """
+        Determine the latest state of the node.
+
+        :param target_node: A libcloud node object representing the compute
+            instance to examine.
+
+        :return: Deferred that fires with a libcloud node object that
+            has the updated information about the node if the node is found,
+            ``None`` otherwise (e.g. if the node has been destroyed).
+        """
+        d = _retry_exception_async(reactor, self._driver.list_nodes)
+
+        def got_nodes(nodes):
+            for node in nodes:
+                if node.uuid == target_node.uuid:
+                    Message.log(
+                        message_type=(
+                            u"flocker:provision:libcloud:refresh_node"
+                        ),
+                        name=node.name,
+                        id=node.id,
+                        state=node.state,
+                        public_ips=node.public_ips,
+                        private_ips=node.private_ips,
+                    )
+                    return node
+            return None
+
+        d.addCallback(got_nodes)
+        return d
+
+    def _async_node_in_state(self, reactor, target_node, target_states):
+        """
+        Determine whether a compute instance is in one of the given states.
+
+        :param target_node: A libcloud node object representing the compute
+            instance to examine.
+        :param set target_states: The states to compare the compute instance's
+            actual state against.
+
+        :return: ``True`` if the compute instance is currently in one of
+            ``target_states``, ``False`` otherwise.
+        """
+        d = self._async_refresh_node(reactor, target_node)
+
+        def check_state(node):
+            if node is not None:
+                if node.state in target_states:
+                    Message.log(
+                        message_type=(
+                            u"flocker:provision:libcloud:node_in_state"
+                        ),
+                        name=node.name,
+                        id=node.id,
+                        state=node.state,
+                    )
+                    return True
+            return False
+
+        d.addCallback(check_state)
+        return d
+
+    def _wait_until_running(self, reactor, node):
+        """
+        Wait until the node is running and its network interface is configured.
+
+        This method fails if the node does not reach the expected state until
+        the predefined timeout expires or if the node goes into an error state
```

Comment:
Yes, I referred to the hardcoded steps as the 'predefined timeout' (in the contrast to a timeout that could have been provided as an argument).